compiler/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -818,25 +818,69 @@ fn try_optimize_array_get_from_previous_set(
     SimplifyResult::None
 }
 
+// If we have an array set whose value is from an array get on the same array at the same index,
+// we can simplify that array set to the array we were looking to perform an array set upon.
+
+// Simple case:
+// v3 = array_get v1, index v2
+// v5 = array_set v1, index v2, value v3
+
+// If we could not immediately simplify the array set from its value, we can try to follow
+// the array set backwards:
+
+// v3 = array_get v1, index v2
+// v5 = array_set v1, index v4, value [Field 100, Field 101, Field 102]
+// v7 = array_set mut v5, index v2, value v3
+
+// We want to optimize `v7` to `v5`. We see that `v3` comes from an array get to `v1`. We follow `v5` backwards and see an array set
+// to `v1` and that array set occurs to a different index. We now know we can simplify `v7` to `v5` as it is unchanged.
```

Comment:
We can't make this optimization safely as we don't have a guarantee that `v2` is different from `v4`. If those are both user inputs then we could have `v2 == v4` and so the final result would be `v1` and returning `v5` would be incorrect.

Reply:
Ah yes 🤦. I will change this optimization to work with just constants and see what we get.